client/asset/eth/multirpc.go

```diff
@@ -0,0 +1,1274 @@
+}
+
+func (m *multiRPCClient) reconfigure(ctx context.Context, settings map[string]string) error {
+	providerDef := settings[providersKey]
+	if len(providerDef) == 0 {
+		return errors.New("no providers specified")
+	}
+	endpoints := strings.Split(providerDef, " ")
+	providers, err := connectProviders(ctx, endpoints, m.log, m.chainID)
+	if err != nil {
+		return err
+	}
+	m.providerMtx.Lock()
+	oldProviders := m.providers
+	m.providers = providers
+	m.endpoints = endpoints
+	m.providerMtx.Unlock()
+	for _, p := range oldProviders {
+		p.ec.Close()
+	}
+	return nil
+}
+
+// cleanReceipts cleans up the receipt cache, deleting any receipts that haven't
+// been access for > receiptCacheExpiration.
+func (m *multiRPCClient) cleanReceipts() {
+	m.receipts.Lock()
+	for txHash, rec := range m.receipts.cache {
+		if time.Since(rec.lastAccess) > receiptCacheExpiration {
+			delete(m.receipts.cache, txHash)
+		}
+	}
+	m.receipts.Unlock()
+}
+
+func (m *multiRPCClient) transactionReceipt(ctx context.Context, txHash common.Hash) (r *types.Receipt, err error) {
+	// Check the cache.
+	const cacheExpiration = time.Minute
+
+	m.receipts.RLock()
+	cached := m.receipts.cache[txHash]
+	if cached != nil && cached.confirmed {
+		cached.lastAccess = time.Now()
+	}
+	if time.Since(m.receipts.lastClean) > time.Minute*20 {
+		m.receipts.lastClean = time.Now()
+		go m.cleanReceipts()
+	}
+	m.receipts.RUnlock()
+
+	// If confirmed or if it was just fetched, return it as is.
+	if cached != nil && (cached.confirmed || time.Since(cached.lastAccess) < cacheExpiration) {
+		return cached.r, nil
+	}
+
+	// Fetch a fresh one.
+	if err = m.withPreferred(func(p *provider) error {
+		r, err = p.ec.TransactionReceipt(ctx, txHash)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+
+	var confs int64
+	if r.BlockNumber != nil {
+		tip, err := m.bestHeader(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("bestHeader error: %v", err)
+		}
+		confs = new(big.Int).Sub(tip.Number, r.BlockNumber).Int64() + 1
+	}
+
+	m.receipts.Lock()
+	m.receipts.cache[txHash] = &receiptRecord{
+		r:          r,
+		lastAccess: time.Now(),
+		confirmed:  confs > txConfsNeededToConfirm,
+	}
+	m.receipts.Unlock()
+
+	return r, nil
+}
+
+type rpcTransaction struct {
+	tx *types.Transaction
+	txExtraDetail
+}
+
+type txExtraDetail struct {
+	BlockNumber *string         `json:"blockNumber,omitempty"`
+	BlockHash   *common.Hash    `json:"blockHash,omitempty"`
+	From        *common.Address `json:"from,omitempty"`
+}
+
+func (tx *rpcTransaction) UnmarshalJSON(b []byte) error {
+	if err := json.Unmarshal(b, &tx.tx); err != nil {
+		return err
+	}
+	return json.Unmarshal(b, &tx.txExtraDetail)
+}
+
+func getRPCTransaction(ctx context.Context, p *provider, txHash common.Hash) (*rpcTransaction, error) {
+	var resp *rpcTransaction
+	err := p.ec.rpc.CallContext(ctx, &resp, "eth_getTransactionByHash", txHash)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, asset.CoinNotFoundError
+	}
```

Comment:
"geth" is pretty ambigious. You probably mean the `ethereum.NotFound` error, which is an `ethclient.Client` interpretation of the `nil` `(*TransactionAPI).GetTransactionByHash` result. But we're using `rpc.Client` directly, since `ethclient.Client` strips out the block and sender details.

```Go
func (ec *Client) TransactionByHash(ctx context.Context, hash common.Hash) (tx *types.Transaction, isPending bool, err error) {
	var json *rpcTransaction
	err = ec.c.CallContext(ctx, &json, "eth_getTransactionByHash", hash)
	if err != nil {
		return nil, false, err
	} else if json == nil {
		return nil, false, ethereum.NotFound
	} else if ...
}
```